src/katsdpmodels/models.py

```diff
@@ -0,0 +1,263 @@
+    Models loaded by the fetcher store the checksum of the raw data, and are
+    considered equal if the checksums match. Otherwise, equality is by object
+    identity.
+
+    Models loaded by the fetcher should not be modified, as they may be shared
+    by other users. Instead, make a copy and modify that.
+
+    Subclassing should generally be done in two layers:
+
+    1. A class that defines `model_type` and defines the interface for that
+       model type. This will be passed to :class:`.Fetcher` to indicate what
+       model type is expected. Due to limitations in mypy, this should not
+       use ``@abstractmethod`` for the interface methods.
+    2. A concrete implementation that defines `model_format`.
+    """
+
+    model_type: ClassVar[str]
+    model_format: ClassVar[str]
+    checksum: Optional[str] = None
+    target: Optional[str] = None
+    comment: Optional[str] = None
+    author: Optional[str] = None
+    created: Optional[datetime] = None
+
+    @classmethod
+    @abstractmethod
+    def from_file(cls: Type[_M], file: io.IOBase, url: str) -> _M:
+        """Load a model from raw data.
+
+        On success, the callee takes responsibility for closing `file`, either
+        within the function itself or the :meth:`close` method of the returned
+        model.
+
+        The `url` may be used to determine the file type.
+        """
+
+    def close(self) -> None:
+        """Close external resources associated with the model.
+
+        Attempting to use the model after that results in undefined behavior.
+        However, it is legal to call :meth:`close` multiple times.
+
+        Models also implement the context manager protocol.
+        """
+
+    def __enter__(self: _M) -> _M:
+        return self
+
+    def __exit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> None:
+        self.close()
+
+    def __eq__(self, other: object) -> Any:
+        if not isinstance(other, Model):
+            return NotImplemented
+        elif self.checksum is not None and other.checksum is not None:
+            return self.checksum == other.checksum and type(self) is type(other)
+        else:
+            return self is other
+
+    def __hash__(self) -> int:
+        if self.checksum is not None:
+            return hash(self.checksum)
+        else:
+            return super().__hash__()
+
+
+class SimpleHDF5Model(Model):
+    """Helper base class for models that load data from HDF5.
+
+    It does not handle lazy loading: the :meth:`from_hdf5` class method must
+    load all the data out of the HDF5 file as it will be closed by the caller.
+    The implementation of :meth:`from_hdf5` does not need to pull out the
+    generic metadata (comment, target, author, created).
+    """
+
+    @classmethod
+    def from_file(cls: Type[_H], file: io.IOBase, url: str) -> _H:
+        with file:
+            parts = urllib.parse.urlparse(url)
+            if not parts.path.endswith(('.h5', '.hdf5')):
```

Comment:
Not sure about this - if the user wants to interpret something random as an HDF5 file, why not let them try. The h5py.File line will error out accordingly if it cannot be parsed.

Reply:
I thought we'd agree that we wanted the flexibility to support other file formats in future, and that we weren't going to depend on the server to supply suitable MIME Content-Type headers; that leaves us with extension to figure out what sort of file we're looking at.

Reply:
Sure, but this is already in an HDF5 class, so aren't you already making assumptions. I just don't see why we shouldn't at least see if whatever we have been given is a valid HDF5 file, regardless of the extension.

Reply:
If we start accepting HDF5 files with the wrong extension now, then they're in the archive forever, and we have to accept them forever, which would mean we can't remove the assumption from the code.

Reply:
I give in, but I really hate file type identification by name.

Reply:
I'm prepared to do it by Content-Type instead. But does RGW have a way to set the Content-Type header per object?

Reply:
It does have some machinery for object type detection. Let's park this for now, but I will test it out and we can potentially add this in a future iteration.

Reply:
Testing with minio at least, one can set a Content-Type header on the upload, and it's remembered and returned when object is retrieved. So we could use Content-Type as the primary decider, falling back to extension if there is no Content-Type or it is application/octet-stream (for example, when reading from a filesystem there is no Content-Type).